control-plane/connect-inject/controllers/pod/pod.go

```diff
@@ -0,0 +1,361 @@
+		Id: getWorkloadID(pod.Name, r.getConsulNamespace(pod.Namespace), r.getPartition()),
+	}
+
+	_, err := r.ConsulResourceServiceClient.Delete(ctx, req)
+	return err
+}
+
+func (r *Controller) deleteHealthStatus(ctx context.Context, pod corev1.Pod) error {
+	return nil
+}
+
+func (r *Controller) writeWorkload(ctx context.Context, pod corev1.Pod) error {
+
+	// TODO: we should add some validation on the required fields here
+	// e.g. what if token automount is disabled and there is not SA. The API call
+	// will fail with no indication to the user other than controller logs
+	ports, workloadPorts := getWorkloadPorts(pod)
+
+	var node corev1.Node
+	// Ignore errors because we don't want failures to block running services.
+	_ = r.Client.Get(context.Background(), types.NamespacedName{Name: pod.Spec.NodeName, Namespace: pod.Namespace}, &node)
+	locality := parseLocality(node)
+
+	workload := &pbcatalog.Workload{
+		Addresses: []*pbcatalog.WorkloadAddress{
+			{Host: pod.Status.PodIP, Ports: ports},
+		},
+		Identity: pod.Spec.ServiceAccountName,
+		Locality: locality,
+		NodeName: common.ConsulNodeNameFromK8sNode(pod.Spec.NodeName),
+		Ports:    workloadPorts,
+	}
+
+	proto, err := anypb.New(workload)
+	if err != nil {
+		return fmt.Errorf("could not serialize workload: %w", err)
+	}
+
+	// TODO: allow custom workload metadata
+	meta := map[string]string{
+		constants.MetaKeyKubeNS: pod.Namespace,
+		metaKeyManagedBy:        constants.ManagedByPodValue,
+	}
+
+	req := &pbresource.WriteRequest{
+		Resource: &pbresource.Resource{
+			Id:       getWorkloadID(pod.GetName(), r.getConsulNamespace(pod.Namespace), r.getPartition()),
+			Metadata: meta,
+			Data:     proto,
+		},
+	}
+	_, err = r.ConsulResourceServiceClient.Write(ctx, req)
+	return err
+}
+
+func (r *Controller) writeHealthStatus(pod corev1.Pod) error {
+	return nil
+}
+
+// TODO(dans): delete ACL token for workload
+// deleteACLTokensForServiceInstance finds the ACL tokens that belongs to the service instance and deletes it from Consul.
+// It will only check for ACL tokens that have been created with the auth method this controller
+// has been configured with and will only delete tokens for the provided podName.
+// func (r *Controller) deleteACLTokensForWorkload(apiClient *api.Client, svc *api.AgentService, k8sNS, podName string) error {
+
+// TODO: add support for explicit upstreams
+//func (r *Controller) writeUpstreams(pod corev1.Pod) error
+
+// shouldIgnore ignores namespaces where we don't mesh-inject.
+func shouldIgnore(namespace string, denySet, allowSet mapset.Set) bool {
+	// Ignores system namespaces.
+	if namespace == metav1.NamespaceSystem || namespace == metav1.NamespacePublic || namespace == "local-path-storage" {
+		return true
+	}
+
+	// Ignores deny list.
+	if denySet.Contains(namespace) {
+		return true
+	}
+
+	// Ignores if not in allow list or allow list is not *.
+	if !allowSet.Contains("*") && !allowSet.Contains(namespace) {
+		return true
+	}
+
+	return false
+}
```

Comment:
~ Move this and the v1 endpoints version to common? Would be nice to consolidate tests for it too